### ISSUE TITLE
Cached aiohttp.web.Response objects cause requests to hang

### ISSUE THREAD

**Issue by lpedrosa** [aio-libs/aiohttp#3020]
## Long story short

If you use a cached `aiohttp.web.Response` in a middleware function it will cause any subsequent requests to hang.

## Expected behaviour

After thinking for a bit, I suggest that we either:

* Improve documentation, by showing an example of what to do if you wish to cache the response body, or
* Make the response object re-usable between requests. It can still remain unchangeable after `prepare()` has been called, preserving its semantics. However, this is probably not feasible with the current architecture.

## Actual behaviour

As far as I understand, `aiohttp.web.Response` is a _Finite State Machine_. Once a `Response`
object is dispatched by the server, it must be emptying some internal buffer. This renders it unusable for subsequent dispatches.

The current solution for avoiding this behaviour seems to be having the middleware generate a new `Response` object, everytime it is called.

## Steps to reproduce

Here is a simple application that reproduces the behaviour:

```python
from aiohttp import web


def my_middleware_factory():

    # this could have been passed in as a default arg, or exist as a global
    state = web.Response(status=200, text='this is cached')

    @web.middleware
    async def middleware(request, handler):
        res = await handler(request)
        return state

    return middleware


async def handler(request):
    return web.Response(status=200, text='Hello World!')


def main():
    app = web.Application()
    app.router.add_get('/', handler)
    app.middlewares.append(my_middleware_factory())

    web.run_app(app)


if __name__ == '__main__':
    main()

```

* When you hit `http://localhost:8080` twice
* Then the second request should hang your http client

## Your environment

Python 3.6
aiohttp=3.2.1
Ubuntu 16.04
'server'

**Comment by lpedrosa** [comment]
I'm glad to help augmenting the current middleware factory usage examples for this. Your framework is really helpful!

I have run into this behaviour while building a generic _Exception to Response_ middleware factory. My current workaround can be found [here](https://github.com/lpedrosa/nccostorage/blob/ab082f0fc9ec2612f62e439c80b1ad977b7f768a/nccostorage/middleware.py)

The `to_response()` function generates a new `web.Response` everytime it is called.

**Comment by asvetlov** [comment]
Yes, web response cannot be used multiple times to answer different requests.
You should recreate it every time, the response object cannot be cached.
The rule of thumb: one request -- one response.

If you want to make a documentation pull request -- you are welcome. I doubt if we need to blow up the main docs but FAQ sounds like a good place for such things.

**Comment by lpedrosa** [comment]
The FAQ sounds like a great place to put it. Let me try to figure out how to do it. 

> The rule of thumb: one request -- one response.

Thanks for clarifying this!

**Comment by asvetlov** [comment]
@lpedrosa ping

**Comment by lpedrosa** [comment]
Hi @asvetlov. Sorry about going MIA. I'll create a PR in a couple of minutes, containing the basic draft of the new FAQ entry.

**Comment by asvetlov** [comment]
Fixed by #3129
